modules/ldapbackend/ldapauthenticator.cc

```diff
@@ -163,98 +190,109 @@ int LdapGssapiAuthenticator::attemptAuth( LDAP *conn )
-    L<<Logger::Error << logPrefix << "krb5 error when initiating keytab search: " << std::string( krb5_get_error_message( context, code ) ) << std::endl;
-    krb5_kt_close( context, keytab );
+  if ( ( code = krb5_kt_start_seq_get( d_context, keytab, &cursor ) ) != 0 ) {
+    L<<Logger::Error << d_logPrefix << "krb5 error when initiating keytab search: " << std::string( krb5_get_error_message( d_context, code ) ) << std::endl;
+    krb5_kt_close( d_context, keytab );
     return code;
   }
 
   krb5_keytab_entry entry;
-  if ( ( code = krb5_kt_next_entry( context, keytab, &entry, &cursor ) ) == 0 ) {
-    code = krb5_copy_principal( context, entry.principal, &principal );
-    krb5_kt_free_entry( context, &entry );
+  if ( ( code = krb5_kt_next_entry( d_context, keytab, &entry, &cursor ) ) != 0 ) {
+    L<<Logger::Error << d_logPrefix << "krb5 error when retrieving first keytab entry: " << std::string( krb5_get_error_message( d_context, code ) ) << std::endl;
+    krb5_kt_close( d_context, keytab );
+    return code;
   }
 
-  krb5_kt_end_seq_get( context, keytab, &cursor );
-  if ( code != 0 ) {
-    L<<Logger::Error << logPrefix << "krb5 error when extracting principal information: " << std::string( krb5_get_error_message( context, code ) ) << std::endl;
-    krb5_kt_close( context, keytab );
-    krb5_free_principal( context, principal );
+  if ( ( code = krb5_copy_principal( d_context, entry.principal, &principal ) ) != 0 ) {
+    L<<Logger::Error << d_logPrefix << "krb5 error when extracting principal information: " << std::string( krb5_get_error_message( d_context, code ) ) << std::endl;
+    krb5_kt_close( d_context, keytab );
+    krb5_kt_free_entry( d_context, &entry );
     return code;
   }
 
-  // Locate the credentials cache file
-  if ( !cCacheFile.empty() ) {
-    std::string cCacheStr( "FILE:" + cCacheFile );
-    code = krb5_cc_resolve( context, cCacheStr.c_str(), &ccache );
+  krb5_kt_free_entry( d_context, &entry );
+  krb5_kt_end_seq_get( d_context, keytab, &cursor );
+
+  if ( ( code = krb5_get_init_creds_opt_alloc( d_context, &options ) ) != 0 ) {
+    L<<Logger::Error << d_logPrefix << "krb5 error when allocating credentials cache structure: " << std::string( krb5_get_error_message( d_context, code ) ) << std::endl;
+    krb5_kt_close( d_context, keytab );
+    krb5_free_principal( d_context, principal );
+    return code;
   }
-  else {
-    code = krb5_cc_default( context, &ccache );
+  krb5_get_init_creds_opt_set_default_flags( d_context, "pdns", krb5_principal_get_realm( d_context, principal ), options );
+
+  // Get the ticket
+  code = krb5_get_init_creds_keytab( d_context, &credentials, principal, keytab, 0, NULL, options );
+  if ( code ) {
+    L<<Logger::Error << d_logPrefix << "krb5 error when getting the TGT: " << std::string( krb5_get_error_message( d_context, code ) ) << std::endl;
+    krb5_free_cred_contents( d_context, &credentials );
+    krb5_kt_close( d_context, keytab );
```

Comment:
We might be missing a call to `krb5_get_init_creds_opt_free( d_context, options );` here.